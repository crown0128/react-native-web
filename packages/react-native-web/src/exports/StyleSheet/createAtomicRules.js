import createRuleBlock from './createRuleBlock';

const createAtomicRules = (selector, prop, value) => {
  const rules = [];

  switch (prop) {
    // pointerEvents is a special case that requires custom values and additional rules
    // See #513
    case 'pointerEvents': {
      let val = value;
      if (value === 'auto' || value === 'box-only') {
        val = 'auto !important';
        if (value === 'box-only') {
          const block = createRuleBlock({ [prop]: 'none' });
          rules.push(`${selector} > *{${block}}`);
        }
      } else if (value === 'none' || value === 'box-none') {
        val = 'none !important';
        if (value === 'box-none') {
          const block = createRuleBlock({ [prop]: 'auto' });
          rules.push(`${selector} > *{${block}}`);
        }
      }
      const block = createRuleBlock({ [prop]: val });
      rules.push(`${selector}{${block}}`);
      break;
    }

    case 'placeholderTextColor': {
      const block = createRuleBlock({ color: value, opacity: 1 });
      rules.push(`@media all {
${selector}::-webkit-input-placeholder{${block}}
${selector}::-moz-placeholder{${block}}
${selector}:-ms-input-placeholder{${block}}
${selector}::placeholder{${block}}
}`);
      break;
    }

    default: {
      const block = createRuleBlock({ [prop]: value });
      rules.push(`${selector}{${block}}`);
    }
  }

  return rules;
};

export default createAtomicRules;
