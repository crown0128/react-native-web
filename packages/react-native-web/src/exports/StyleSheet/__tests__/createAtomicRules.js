/* eslint-env jasmine, jest */

import createAtomicRules from '../createAtomicRules';

describe('StyleSheet/createAtomicRules', () => {
  test('transforms standard declarations to a single rule', () => {
    expect(createAtomicRules('.test', 'margin', 0)).toMatchSnapshot();
  });

  test('transforms custom pointerEvents declaration', () => {
    expect(createAtomicRules('.test', 'pointerEvents', 'box-only')).toMatchSnapshot();
  });

  test('transforms custom placeholderTextColor declaration', () => {
    expect(createAtomicRules('.test', 'placeholderTextColor', 'gray')).toMatchSnapshot();
  });
});
