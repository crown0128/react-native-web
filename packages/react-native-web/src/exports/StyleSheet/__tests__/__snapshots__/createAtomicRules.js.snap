// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheet/createAtomicRules transforms custom placeholderTextColor declaration 1`] = `
Array [
  "@media all {
.test::-webkit-input-placeholder{color:gray;opacity:1}
.test::-moz-placeholder{color:gray;opacity:1}
.test:-ms-input-placeholder{color:gray;opacity:1}
.test::placeholder{color:gray;opacity:1}
}",
]
`;

exports[`StyleSheet/createAtomicRules transforms custom pointerEvents declaration 1`] = `
Array [
  ".test > *{pointer-events:none}",
  ".test{pointer-events:auto !important}",
]
`;

exports[`StyleSheet/createAtomicRules transforms standard declarations to a single rule 1`] = `
Array [
  ".test{margin:0px}",
]
`;
